src/test/run-pass/async-fn-size.rs

```diff
@@ -0,0 +1,106 @@
+// edition:2018
+// aux-build:arc_wake.rs
+
+#![feature(async_await, await_macro)]
+
+extern crate arc_wake;
+
+use std::pin::Pin;
+use std::future::Future;
+use std::sync::{
+    Arc,
+    atomic::{self, AtomicUsize},
+};
+use std::task::{Context, Poll};
+use arc_wake::ArcWake;
+
+struct Counter {
+    wakes: AtomicUsize,
+}
+
+impl ArcWake for Counter {
+    fn wake(self: Arc<Self>) {
+        Self::wake_by_ref(&self)
+    }
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        arc_self.wakes.fetch_add(1, atomic::Ordering::SeqCst);
+    }
+}
+
+struct WakeOnceThenComplete(bool, u8);
+
+impl Future for WakeOnceThenComplete {
+    type Output = u8;
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<u8> {
+        if self.0 {
+            Poll::Ready(self.1)
+        } else {
+            cx.waker().wake_by_ref();
+            self.0 = true;
+            Poll::Pending
+        }
+    }
+}
+
+fn wait(fut: impl Future<Output = u8>) -> u8 {
+    let mut fut = Box::pin(fut);
+    let counter = Arc::new(Counter { wakes: AtomicUsize::new(0) });
+    let waker = ArcWake::into_waker(counter.clone());
+    let mut cx = Context::from_waker(&waker);
+    loop {
+        match fut.as_mut().poll(&mut cx) {
+            Poll::Ready(out) => return out,
+            Poll::Pending => (),
+        }
+    }
+}
+
+fn base() -> WakeOnceThenComplete { WakeOnceThenComplete(false, 1) }
+
+async fn foo1() -> u8 {
+    await!(base())
+}
+
+async fn foo2() -> u8 {
+    await!(base()) + await!(base())
+}
+
+async fn foo3() -> u8 {
+    await!(base()) + await!(base()) + await!(base())
+}
+
+async fn bar() -> u8 {
+    await!(foo3()) + await!(foo3()) + await!(foo3())
+}
+
+async fn baz() -> u8 {
+    await!(bar()) + await!(bar()) + await!(bar())
+}
+
+async fn bazz() -> u8 {
+    await!(baz()) + await!(baz()) + await!(baz())
+}
+
+async fn bazzz() -> u8 {
+    await!(bazz()) + await!(bazz()) + await!(bazz())
+}
+
+fn main() {
```

Comment:
Originally, these were the sizes:

```
2    base()
8    await1_level1()
12   await2_level1()
12   await3_level1()
44   await3_level2()
140  await3_level3()
428  await3_level4()
1292 await3_level5()
```

This does increase size of `await1_level1()` by 4 bytes, because of the "prefix" trick I'm using in the layout computation. We can do better with a greedy placement algorithm, but representing the state of that algorithm is quite tricky. This prevents future sizes from multiplying each time you add a new level of `await` nesting, and I think that's much more important right now.